system/lib/wasmfs/syscalls.cpp

```diff
@@ -663,54 +632,49 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
 enum class UnlinkMode { Rmdir, Unlink };
 
 static long doUnlink(char* path, UnlinkMode unlinkMode) {
-  auto pathParts = splitPath(path);
-
-  // TODO: Ensure that . and .. are invalid when path parsing is updated.
-  // This can be done when path parsing is refactored.
-
-  long err;
-  auto parsedPath = getParsedPath(pathParts, err);
-
-  // Parent node doesn't exist.
-  if (!parsedPath.parent) {
+  auto parsed = path::parseParent(path);
+  if (auto err = parsed.getError()) {
     return err;
   }
-
-  if (!parsedPath.child) {
+  auto& [parent, childNameView] = parsed.getParentChild();
+  std::string childName(childNameView);
+  if (childName == ".") {
```

Comment:
why is this case special? I think it means this is a directory - can we not return ISDIR here?

Reply:
Oh yeah, I guess this should be `EISDIR` for `unlink` but `EINVAL` for `rmdir`. `unlinkat` (not implemented yet) actually changes its behavior to match one or the other depending on whether it is passed `AT_REMOVEDIR`, which simplifies things.

Reply:
Can we remove the check of `childName == "."`, though? Isn't it covered already by the other logic?

Reply:
Wthout this check `rmdir("/foo/.")` would incorrectly succeed, but you're right that this is already covered in the `UnlinkMode::Unlink` case. I'll make this check conditional on having `UnlinkMode::Rmdir`.

Reply:
There was a bug here, which was that parsing "/" would trigger this error because it makes `parseParent` return `(root, ".")`. To fix this, I had to insert logic to manually find and check the last segment (see my last commit). Maybe it would be better to skip that extra logic and just accept returning EINVAL instead of the correct EBUSY in that case?